cmd/bcpd/app/app_test.go

```diff
@@ -114,163 +110,196 @@ func TestApp(t *testing.T) {
+		namecoin.Wallet{
+			Name: "demote",
+			Coins: x.Coins{
+				{
+					Ticker: "ETH",
+					Whole:  48000,
+				},
+				{
+					Ticker: "FRNK",
+					Whole:  1234,
+				},
+			},
+		})
 
 	// get a token
-	tquery := abci.RequestQuery{
-		Path: "/tokens",
-		Data: []byte("ETH"),
-	}
-	var toke namecoin.Token
-	tres := myApp.Query(tquery)
-	err = app.UnmarshalOneResult(tres.Value, &toke)
-	require.NoError(t, err)
-	assert.Equal(t, int32(9), toke.SigFigs)
-	assert.Equal(t, "Smells like ethereum", toke.Name)
+	queryAndCheckToken(t, myApp, "/tokens", []byte("ETH"),
+		[]namecoin.Token{
+			{
+				Name:    "Smells like ethereum",
+				SigFigs: int32(9),
+			},
+		})
 
 	// get all tokens
-	aquery := abci.RequestQuery{
-		Path: "/tokens?prefix",
-	}
-	ares := myApp.Query(aquery)
-	var set app.ResultSet
-	err = set.Unmarshal(ares.Value)
-	require.NoError(t, err)
-	assert.Equal(t, 2, len(set.Results))
-	err = toke.Unmarshal(set.Results[1])
-	require.NoError(t, err)
-	assert.Equal(t, int32(3), toke.SigFigs)
-	assert.Equal(t, "Frankie", toke.Name)
-
-	// testing multisig contract
-	// first create a contract
-	// then a wallet at the contract address
-	// finaly send money from the wallet controlled by contractAddr
-
-	// create contract
-	ck1 := crypto.GenPrivKeyEd25519()
-	ck2 := crypto.GenPrivKeyEd25519()
-	ck3 := crypto.GenPrivKeyEd25519()
-	signers := [][]byte{
-		ck1.PublicKey().Address(),
-		ck2.PublicKey().Address(),
-		ck3.PublicKey().Address()}
-	cmsg := &multisig.CreateContractMsg{
-		Sigs:                signers,
-		ActivationThreshold: 2,
-		AdminThreshold:      2,
-	}
-	tx = &Tx{
-		Sum: &Tx_CreateContractMsg{cmsg},
-	}
-	dres = signAndCommit(t, myApp, tx, []Signer{{pk, 1}}, chainID, 3)
+	queryAndCheckToken(t, myApp, "/tokens?prefix", nil,
+		[]namecoin.Token{
+			{
+				Name:    "Smells like ethereum",
+				SigFigs: int32(9),
+			},
+			{
+				Name:    "Frankie",
+				SigFigs: int32(3),
+			},
+		})
+
+	// create recoveryContract
+	recovery1 := crypto.GenPrivKeyEd25519()
+	recovery2 := crypto.GenPrivKeyEd25519()
+	recovery3 := crypto.GenPrivKeyEd25519()
+	recoveryContract := createContract(t, myApp, chainID, 3, []Signer{{pk, 1}},
+		2, recovery1.PublicKey().Address(), recovery2.PublicKey().Address(), recovery3.PublicKey().Address())
+
+	// create safeKeyContract contract
+	// can be activated by masterKey or recoveryContract
+	masterKey := crypto.GenPrivKeyEd25519()
+	safeKeyContract := createContract(t, myApp, chainID, 4, []Signer{{pk, 2}},
+		1, masterKey.PublicKey().Address(), multisig.MultiSigCondition(recoveryContract).Address())
+
+	// create a wallet controlled by safeKeyContract
+	safeKeyContractAddr := multisig.MultiSigCondition(safeKeyContract).Address()
+	sendToken(t, myApp, chainID, 5, []Signer{{pk, 3}},
+		addr, safeKeyContractAddr, 2000, "ETH", "New wallet controlled by safeKeyContract")
+
+	// build and sign a transaction using master key to activate safeKeyContract
```

Comment:
Nice workflow leading up to these two examples.
Suggestion to make a separate test case than TestApp... we can have TestSend, TestMultiSig, TestSwap, etc... but for now fine.